doc/keynames/README.md

```diff
@@ -0,0 +1,477 @@
+In fact, we will see [later](#4-valid-and-invalid-key-names), that removing the `/` makes the Key Name invalid.
+
+There also is a completely new addition in Elektra.
+Elektra has a notion of _Array Parts_.
+These are Key Name Parts that denote an array index.
+How exactly these work, will be explored [later](#4-valid-and-invalid-key-names).
+For now, we only need to know that they start with an `#` and their canonical form has `n` underscores followed by `n+1` digits.
+
+A few examples for Array Parts:
+
+| Non-Canonical    | Canonical           |
+| ---------------- | ------------------- |
+| `/elektra/#10`   | `/elektra/#_10`     |
+| `/elektra/#1234` | `/elektra/#___1234` |
+
+### 2.2. Other Escape Sequences
+
+We already know, that `/` and `\` have to be escaped in an Escaped Key Name.
+In addition two these to, there are a few more characters that have to be escaped.
+However, these additional characters may **only** be escaped at the start of a Key Name Part.
+
+The characters in question are: `.`, `#`, `%`, `@`.
+For example the Canonical Escaped Key Name for the Key with Namespace "Cascading" and Key Name Parts `elektra`, `.` and `version` is: `/elektra/\./version`.
+The `\.` is the escaped form of `.`.
+Using just `.` would make it a Non-Canonical Key Name, whose Canonical counterpart is `/elektra/version`, which does not have the wanted Key Name Part `.`.
+Similarly, `#10` would make a Key Name Non-Canonical, but `\#10` won't.
+Therefore, if you want a Canonical Key Name with a Key Name Part `#10` and not actually `#_10`,
+must use e.g. `/elektra/\#10/version`.
+
+<!-- TODO (kodebach): Unforunate escaping behavior
+        Unfortunately this means that both `/elektra/.version` and `/elektra/\.version` have the same unescaped form (`elektra`, `.version`).
+        The same applies to `/elektra/%version` and `/elektra/\%version`.
+        All of these are, however, Canonical Key Names.
+        So the Key (CASCADING, `elektra`, `.version`) actually has two canonical forms....
+-->
+
+## 3. Unescaped Names
+
+While the Escaped Name of a Key is nice for humans, it is not very well suited for machines.
+The escaping of path separators makes it hard to find the Key Name Parts of a given Key Name.
+The Escaped Name is also not well suited for sorting [[3]](#footnote-3).
+<a id="ref-footnote-3"></a>
+
+Both of these flaws are solved by the _Unescaped Name_.
+In Unescaped Names we use a zero-byte as the Part Separator.
+Since a Key Name Part cannot contain a zero-byte, we do not need an escaping mechanism for the path separator.
+
+However, these zero-bytes mean that the Unescaped Name is not a printable string and therefore not human-readable.
+This why we will only describe the Unescaped Name in the deconstructed form.
+
+The Escaped Names from above correspond to the following Unescaped Names:
+
+1. `system:/elektra/version/info`
+   - The byte representing the Namespace "System": `0x07`
```

Comment:
That would be confusing IMO. Then you would have to refer to the header file to find out what the enum entry actually means. It also doesn't have an advantage, to only use the enum entries. We can't easily change the values anyway, because that would impact the sort order of a KeySet.

But just in case, I will add a note stating that the byte values may have changed and that the `enum` in `kdb.h.in.c` is the source of truth